microsoft/onefuzz
# Fix notifications bugs

## Summary of the Pull Request

_What is this about?_

Implements 1. and 3. from #2530 

## PR Checklist

## Info on Pull Request

_What does this include?_

## Validation Steps Performed

1. Create a work item through `onefuzz debug notification job {jobId}`. This validates this block:
```csharp
               // We never saw a work item like this before, it must be new
              else {
                var entry = await CreateNew();
                var adoEventType = "AdoNewItem";
                _logTracer.WithTags(notificationInfo).Event($"{adoEventType} {entry.Id:Tag:WorkItemId}");
            }
```
   Verify the work item was created properly.
2. Re-run `onefuzz debug notification job {jobId}` to make sure the same work item can continue to be updated. This validates this block:
```csharp
            } else if (numNonDuplicate == 1) {
                await UpdateExisting(await nonDuplicate.SingleAsync(), notificationInfo);
            }
```
   Verify the work item was updated properly.
3. Manually create a duplicate of the work item in ADO then re-run `onefuzz debug notification job {jobId}`. This will validate having multiple work items that aren't marked as duplicate in this block:
```csharp
            if (numNonDuplicate > 1) {
                var nonDuplicateWorkItemIds = nonDuplicate
                    .Select(wi => wi.Id)
                    .ToEnumerable();


                var matchingWorkItemIds = matchingWorkItems
                    .Select(wi => wi.Id)
                    .ToEnumerable();


                var extraTags = new List<(string, string)> {
                    ("NonDuplicateWorkItemIds", JsonSerializer.Serialize(nonDuplicateWorkItemIds)),
                    ("MatchingWorkItemIds", JsonSerializer.Serialize(matchingWorkItemIds))
                };
                extraTags.AddRange(notificationInfo);


                _logTracer.WithTags(extraTags).Info($"Found more than 1 matching, non-duplicate work item");
                await foreach (var workItem in nonDuplicate) {
                    await UpdateExisting(workItem, notificationInfo);
                }
```
   Verify both work items were updated properly and log message created.
4. Mark the manually copied work item from Step 3 as "Duplicate" using the "Related Work" panel in the work item.
5. Re-run `onefuzz debug notification job {jobId}`, this will validate this block:
```csharp
            var matchingWorkItems = ExistingWorkItems();


            var nonDuplicate = matchingWorkItems
                .Where(wi => !IsDuplicate(wi));


            var numNonDuplicate = await nonDuplicate.CountAsync();
```
   Validate that only the work item created in step 1 was updated.
6. Mark the work item created in step 1 as "Resolved" with reason "Duplicate", this will validate this block:
```csharp
              // We have matching work items but all are duplicates
              else if (await matchingWorkItems.AnyAsync()) {
                _logTracer.WithTags(notificationInfo).Info($"All matching work items were duplicates, re-opening the oldest one");
                var oldestWorkItem = await matchingWorkItems.OrderBy(wi => wi.Id).FirstAsync();
                await UpdateExisting(oldestWorkItem, notificationInfo);


                _ = await _client.AddCommentAsync(
                    new CommentCreate() {
                        Text = "This work item was re-opened because OneFuzz could only find related work items that are marked as duplicate."
                    },
                    _project,
                    (int)oldestWorkItem.Id!);
            }
```
   Validate that the work item created in step 1 is re-opened, is updated, and has a new comment explaining why it was re-opened.
